utils/visualizer.py

```diff
@@ -49,16 +52,468 @@ def __convert_to_list(self):
         self.true_values = list(chain.from_iterable(self.true_values))
         self.predicted_values = list(chain.from_iterable(self.predicted_values))
 
-    def create_scatter_plot(self, save_plot=True):
+    def hist2d_contour(self, data1, data2):
+        hist2d_pasr, xedge_pasr, yedge_pasr = np.histogram2d(
+            np.hstack(data1), np.hstack(data2), bins=50
+        )
+        xcen_pasr = 0.5 * (xedge_pasr[0:-1] + xedge_pasr[1:])
+        ycen_pasr = 0.5 * (yedge_pasr[0:-1] + yedge_pasr[1:])
+        BCTY_pasr, BCTX_pasr = np.meshgrid(ycen_pasr, xcen_pasr)
+        hist2d_pasr = hist2d_pasr / np.amax(hist2d_pasr)
+        return BCTX_pasr, BCTY_pasr, hist2d_pasr
+
+    def hist1d_err(self, data1, data2, weight=1.0):
+        errabs = np.abs(np.hstack(data1) - np.hstack(data2)) * weight
+
+        hist2d_pasr, xedge_pasr, yedge_pasr = np.histogram2d(
+            np.hstack(data1), errabs, bins=50
+        )
+        xcen_pasr = 0.5 * (xedge_pasr[0:-1] + xedge_pasr[1:])
+        ycen_pasr = 0.5 * (yedge_pasr[0:-1] + yedge_pasr[1:])
+        hist2d_pasr = hist2d_pasr / np.amax(hist2d_pasr)
+        mean1d_cond = np.dot(hist2d_pasr, ycen_pasr) / np.sum(hist2d_pasr, axis=1)
+        return xcen_pasr, mean1d_cond
+
+    def create_scatter_plot(self, ivar, save_plot=True):
         """Creates scatter plot from stored values in the tru and  predicted values lists."""
-        self.__convert_to_list()
-        plt.scatter(self.true_values, self.predicted_values)
-        plt.title("Scatter plot for true and predicted values.")
-        plt.xlabel("True value")
-        plt.ylabel("Predicted value")
+        # self.__convert_to_list()
+        nshape = np.asarray(self.predicted_values).shape
+        if nshape[1] == 1:
+            fig = plt.figure(figsize=(15, 4.5))
+            plt.subplots_adjust(
+                left=0.08, bottom=0.15, right=0.95, top=0.925, wspace=0.35, hspace=0.1
+            )
+            ax = plt.subplot(1, 3, 1)
+            plt.scatter(self.true_values, self.predicted_values)
```

Comment:
You could create a new function like `__scatter_impl(x, y,  title, xlabel, ylabel, xlim, ylim)`. Then a similar `__plot_impl(x, y,  title, xlabel, ylabel, xlim, ylim)` and reuse them throughout the class

Reply:
Updated in 66c3241